src/globus_sdk/services/auth/consents.py

```diff
@@ -0,0 +1,311 @@
+            requirement, including dependent scope requirements, is satisfied by the
+            tree.
+    * ``ConsentForest``
+        * A collection of all ConsentTrees for a user rooted under a particular client
+            (the client that initiated the request for consents).
+        * A `meets_scope_requirements` method is defined to check whether a scope
+            requirement, including dependent scope requirements, is satisfied by any
+            tree in the forest.
+"""
+
+from __future__ import annotations
+
+import json
+import typing as t
+from dataclasses import dataclass
+from datetime import datetime
+
+from globus_sdk import Scope
+from globus_sdk._types import UUIDLike
+
+from .errors import ConsentParseError
+
+
+@dataclass
+class Consent:
+    """
+    Consent Data Object
+
+    This object models:
+        * A grant which a user has provided for a client to perform a particular
+            scoped operation on their behalf.
+        * The consent is conditional on the invocation path leading to the client's
+            attempted operation being initiated through a chain of similarly scoped
+            operations (consents) defined in the "dependency_path".
+    """
+
+    client: UUIDLike
+    scope: UUIDLike
+    scope_name: str
+    id: int
+    effective_identity: UUIDLike
+    # A list representing the path of consent dependencies lading from a "root consent"
+    #   to this. The last element of this list will always be this consent's ID.
+    # Downstream dependency relationships may exist but will not be defined here.
+    dependency_path: list[int]
+    created: datetime
+    updated: datetime
+    last_used: datetime
+    status: str
+    allows_refresh: bool
+    auto_approved: bool
+    atomically_revocable: bool
+
+    @classmethod
+    def load(cls, data: t.Mapping[str, t.Any]) -> Consent:
+        try:
+            return cls(
+                id=data["id"],
+                client=data["client"],
+                scope=data["scope"],
+                effective_identity=data["effective_identity"],
+                dependency_path=data["dependency_path"],
+                scope_name=data["scope_name"],
+                created=datetime.fromisoformat(data["created"]),
+                updated=datetime.fromisoformat(data["updated"]),
+                last_used=datetime.fromisoformat(data["last_used"]),
+                status=data["status"],
+                allows_refresh=data["allows_refresh"],
+                auto_approved=data["auto_approved"],
+                atomically_revocable=data["atomically_revocable"],
+            )
+        except KeyError as e:
+            raise ConsentParseError(
+                f"Failed to load Consent object. Missing required key: {e}. "
+                f"Raw Consent: {json.dumps(data)}."
+            ) from e
+
+    def __str__(self) -> str:
+        client = f"{str(self.client)[:8]}..."
```

Comment:
Screen width. Trying to make it so that every serialized consent doesn't require a line wrap.

Reply:
Here's a str-ified consent object for comparison

With Truncation
```
'Consent [433892]: Client [cf01eb30...] -> Scope [https://auth.globus.org/scopes/7707a06d-535c-41fe-bf2b-b69294c334ae/data_access]'
```

Without Truncation
```
'Consent [433892]: Client [cf01eb30-9884-11e5-8d77-87f1f8b059db] -> Scope [https://auth.globus.org/scopes/7707a06d-535c-41fe-bf2b-b69294c334ae/data_access]'
```

Reply:
I think we should go with the full content, if that's the only reason.
`__str__` for objects offers no guarantee that it will be brief, summary, or fit in some fixed size. It's just supposed to be "some sensible string representation of the object".

This goes in multiple directions depending on the data (and author) in question.
e.g. `<Response [200]>` from `requests`, but a full splatted-out nested structure if you `str(some_big_dict)`.

Since we're already including scope strings here, it seems like philosophically we're trying to provide a detailed view of the data, as opposed to the ultra-summary `Consent[433892]`. In that case, I think we're closer to the model of `str(some_dict)` and we should show all of the details.

cc @kurtmckee, I'd really like to get your thoughts or input on this. I'm curious if you have a different perspective about how `__str__` ought to behave. Note that we have another thread in this PR about the `__str__` of an object, in which I'm advocating for producing a `__str__` which may be quite sizeable.

I think it's fine for `__str__` to be huge, but `__repr__` should try to be a compact but totally unambiguous representation -- which means `__repr__` ought not to contain much excess whitespace, but can be very long. `__str__`, on the other hand, I think can contain newlines or any other whitespace which makes data easy to understand.
For an example of my preference on this front in action in the SDK, here's a large `__str__` which I wrote:
https://github.com/globus/globus-sdk-python/blob/7479975d9422d71992df0ebbbfb8ac31daba66ff/src/globus_sdk/scopes/builder.py#L157-L160

Reply:
Not trying to guarantee briefness; `__str__` is a largely judgement-call based interface (as opposed to `__repr__` which has a lot more expectations about what is and isn't included).

I think you're posing a bit of a false dichotomy here that `__str__` should be either (1) ultra summary or (2) entirely comprehensive.
For this object, there's too much information to be entirely comprehensive (i don't want to include, for instance, "updated" in this string) but the ultra summary version (`Consent[433892]`) doesn't provide sufficient information for understanding what the consent represents (I think we agree on this point).

---

With that all in mind, it's a question of balancing  the string representation comprehensibility with the amount of information we're providing. My contention is that:
 1.  consent id, scope string, and associated client are the only pieces relevant to be included in a string representation (I think we still agree up to this point). 
 2. the full client id provides the same amount of information as a truncated client id because this client id won't be used for lookup but rather as a human readable identifier. Ideally we'd have an actually human readable identifier for these clients "transfer" or "dereks dev client" but given that we don't the first n characters of a client id serves the same purpose as the full client id. If that's true, then putting the full client id just elongates the string making parsing the whole thing harder for humans.

Reply:
Okay; I think we're moving a good direction. I agree that it's not fully a choice between "ultra-summary" and "full info" -- rather, I wanted to point out that there's prior art in both directions, so we're free to use quite a lot of discretion for `__str__`.

I also agree that we can wish for "nice client names" as part of the API, but acknowledge that those aren't present today.

Here's where we probably disagree:
> this client id won't be used for lookup but rather as a human readable identifier

I expect that the client ID _will_ be used for lookup. The use-case here is that a user fetched some consent information, and then wanted to print it out and see what all of it was.

At that point, if I have a partial client ID, I think that's almost-useless information for me, as a user. I can't do a lookup with it, which is one of the ways I might want to try to make sense of the information -- e.g. feed it into `globus get-identities -v $CLIENT_ID`.

So I think that _if_ the Client ID is one of the relevant fields for this view (and, like you said, we seem to be in sync on that), then the useful info is the full ID.

Reply:
@sirosen My perspective is "there is no 'ought to'" for `__str__` or `__repr__`. They exist simply as default serialization methods that the language calls automatically.

The implementation here makes a concrete decision that composes three values into a string, accompanied by symbols and nouns in English. If someone else needs the output in a different language, or wants wrapping characters other than `[]`, or wants to align those values into columns in a table, or wants other values altogether, they're free to write their own f-strings to accomplish that.

@derek-globus I only recommend considering if this implementation is anticipating the most common use case.

Reply:
De-truncated client id.